webgl: premultiplication, y flipping, and format conversion

anholt opened the pull request:
<!-- Please describe your changes on the following line: -->
This series implements a bunch of the texture unpack path features for WebGL.  There are known issues with big-endian systems noted in the comments, but I don't know of a clean way to cast from `Vec<u16>` to `Vec<u8>` (which, if we had one, would make this code much cleaner anyway).

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/15122)
<!-- Reviewable:end -->

anholt committed:
webgl: Add support for premultiplying alpha.

Now the affected testcase only fails due to unpack alignment.

anholt committed:
webgl: Convert non-raw TexImage sources to the requested format.

The code was returning RGBA8 data from the non-raw sources (HTML
canvas elements, JS ImageData, etc.), but we then validated and passed
that rgba8 data as if it was whatever format/datatype was specified in
TexImage2D/TexSubImage2D, so the pixels would come out as garbage.

It would seem like we could just rewrite the passed in format/datatype
for the TexImage call to be RGBA/UNSIGNED_BYTE, but that would leave
incorrect levels of precision if the internalformat didn't match the
format/datatype (and older desktop implementations often ignore the
internalformat in choosing their internal format, anyway).

anholt committed:
webgl: Implement FLIP_Y_AXIS unpacking flag for TexImage, TexSubImage.

This fixes a couple of tests doing RGBA/ubyte image uploads with
flipping.  Other tests with flipping get their expectations changed,
because some other feature is missing (premultiplication or
ImageData/canvas format conversion)

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webglrenderingcontext.rs
 * @fitzgen: components/script/dom/webglrenderingcontext.rs
 * @emilio: components/script/dom/webglrenderingcontext.rs

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -471,6 +545,99 @@ impl WebGLRenderingContext {
+    fn premultiply_pixels(&self,
+                          format: TexFormat,
+                          data_type: TexDataType,
+                          pixels: Vec<u8>) -> Vec<u8> {
```
Just as above, this only returns a vector of the same size as `pixels`, right?

If that's the case, probably we should also premultiply in-place, either using `&mut [u8]`, or `mut pixels: Vec<u8>`.

Mind at least mentioning it here (and filing a followup maybe?)

emilio commented on tests/wpt/metadata/webgl/conformance-1.0.3/conformance/textures/tex-image-and-sub-image-2d-with-array-buffer-view.html.ini:
```diff
@@ -1,2171 +1,122 @@
-  [WebGL test #1095: at (0, 4) expected: 0,0,0,255 was 0,0,255,255]
-    expected: FAIL
-
-  [WebGL test #1097: at (0, 8) expected: 0,0,0,255 w
```
Wow, poor test, I was wondering where the diffstat came from, awesome! :)

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -112,6 +112,11 @@ fn has_invalid_blend_constants(arg1: u32, arg2: u32) -> bool {
     }
 }
+
+fn multiply_u8_pixel(a: u8, b: u8) -> u8 {
```
nit: We have code for premultiplication in `components/canvas_traits/lib.rs`. They contain code similar to this, you may consider putting it there (or even all the premultiplication subroutines? Though that's not great because they aren't used anywhere else).

Anyway, your choice :)

anholt replied:
I had noticed that, but I didn't imagine that that code would want to grow some sort of format enum for handling different GL formats.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -398,10 +474,8 @@ impl WebGLRenderingContext {
-                // is coherent with the image.
-                //
-                // RGB8 should be easy to support too
+                // TODO(anholt): What other incoming formats do we
```
I think for now Servo stores all its images internally as RGBA8, so we don't need to worry about it I think.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -471,6 +545,99 @@ impl WebGLRenderingContext {
+                premul
+            }
+
+            _ => pixels
```
Is this also unreachable? If not, consider commenting which cases do we expect? Otherwise, mark as such?

emilio reviewed (approved):
Looks great! Only minor comments, thanks for doing this!

emilio commented:
Regarding casting `Vec`s, you probably can do something using `slice::from_raw_parts{_mut}`, but I think we should only introduce unsafe code if totally necessary. Probably the conversion is not going to be more costly than the IPC message, but if we want to optimize it we could use that I think.

anholt commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -471,6 +545,99 @@ impl WebGLRenderingContext {
+                       internal_format: TexFormat,
+                       data_type: TexDataType,
+                       width: usize,
+                       height: usize) -> Vec<u8> {
```
We're passed in a non-mut Vec at the top of the tex_image()s since we shouldn't be rewriting the JS's array.  We could potentially see that we're going to be chaining more than one of these fixups and use a single mut Vec that we operate on, but that sounded messy.

anholt commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -364,6 +369,77 @@ impl WebGLRenderingContext {
+            // into an enum yet so there's a default case here.  All
+            // WebGL types are 4 bytes per pixel or smaller, so just
+            // return the incoming RGBA8 pixels in this path.
+            _ => pixels
```
OK, looks like unreachable!() should always panic.  I thought I'd seen something in my other debugging that suggested that it didn't panic.

Ms2ger commented:
Perhaps https://docs.rs/byteorder/1.0.0/byteorder/ would help.

emilio reviewed (approved):
r=me, thanks Eric :)

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 41a4f63 has been approved by `emilio`

<!-- @bors-servo r=emilio 41a4f63eb77ef4dfe3bca129f4201ac5816725d2 -->

bors-servo commented:
:hourglass: Testing commit 41a4f63 with merge 82e854a...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/2028)

anholt commented:
OK, I think I know the cause of that Mac failure: not having unpack alignment support.

emilio commented:
We could mark those as failing on mac if you want, I think this is worth landing

anholt committed:
webgl: Fix handling of UNPACK_ALIGNMENT.

We were setting it to whatever value from {1,2,4,8} the user requested
and otherwise ignoring it.  There were two problems there:

1) Validation ignored it, so GL could read outside of the user's array
   in TexImage() or TexSubImage() if the aligment was greater than
   cpp.

2) TexImage()/TexSubImage() from image/canvas sources wasn't packing
   its data according to the unpack alignment.

To fix this, start tracking the user-requested alignment in the DOM
side of the context.  Set the GL's alignment to 1 for image/canvas
sources or the user's value for array sources, and pass the user's
alignment in to validation so that it can figure out the correct size
of image that the GL will ready.

anholt commented:
I'm actually testing the unpack fix now.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -527,8 +530,17 @@ impl WebGLRenderingContext {
-        let expected_byte_length = width * height * element_size * components / components_per_element;
-        return Ok(expected_byte_length);
+        if height == 0 {
+            return Ok(0);
```
nit: Since you early-return, there's no need for the else branch, but not a big deal.

emilio reviewed (commented):
As a uber-nit, `s/ready.$/read./` in the last commit message, but I don't really want to block landing this on it.

emilio commented:
@bors-servo r+

Thanks!

bors-servo commented:
:pushpin: Commit fcef92f has been approved by `emilio`

<!-- @bors-servo r=emilio fcef92f5ba308af3ab6b1ebafd6425995fabe834 -->

bors-servo commented:
:hourglass: Testing commit fcef92f with merge 536c0d7...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/5192), [arm32](http://build.servo.org/builders/arm32/builds/5380), [arm64](http://build.servo.org/builders/arm64/builds/5420), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5523), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/2205), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/2221), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/5502), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5537), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/2194), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/2251), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/662), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/666)
Approved by: emilio
Pushing 536c0d74994bf8b5eaef3f64247753d75ba7baab to master...
